src/core/DrumSynth.cpp

```diff
@@ -172,37 +171,19 @@ void DrumSynth::GetEnv(int env, const char* sec, const char* key, QString ini)
 
 float DrumSynth::waveform(float ph, int form)
 {
-	float w;
-
-	switch (form)
-	{
-	case 0:
-		w = static_cast<float>(sin(fmod(ph, TwoPi)));
-		break; // sine
-	case 1:
-		w = static_cast<float>(fabs(2.0f * static_cast<float>(sin(fmod(0.5f * ph, TwoPi))) - 1.f));
-		break; // sine^2
-	case 2:
-		while (ph < TwoPi)
-		{
-			ph += TwoPi;
-		}
-		w = 0.6366197f * static_cast<float>(fmod(ph, TwoPi) - 1.f); // tri
-		if (w > 1.f)
-		{
-			w = 2.f - w;
-		}
-		break;
-	case 3:
-		w = ph - TwoPi * static_cast<float>(static_cast<int>(ph / TwoPi)); // saw
-		w = (0.3183098f * w) - 1.f;
-		break;
-	default:
-		w = (sin(fmod(ph, TwoPi)) > 0.0) ? 1.f : -1.f;
-		break; // square
-	}
-
-	return w;
+	// sine
+	if (form == 0) { return std::sin(ph); }
+	// sine^2
+	if (form == 1) { return 2.f * std::abs(std::sin(0.5f * ph)) - 1.f; }
```

Comment:
```suggestion
	if (form == 1) { return std::abs(2.f * std::sin(0.5f * ph) - 1.f); }
```
If ph==0, your version would output -1 while the old function would output +1.

Here's what I've been using to test my functions:
https://gist.github.com/messmerd/5ee2e0078b574b9878bdd7224192531e

I've added your functions there too for comparison

Reply:
> If ph==0, your version would output -1 while the old function would output +1.

Yes, and the old function wasn't supposed to. It's an error introduced in b2f2fc4; see the graphs of the functions: https://github.com/LMMS/lmms/pull/7685#discussion_r1942196103

Consider the original code prior to b2f2fc4:

```c++
// original code
case 1: w = (float)fabs(2.0f*(float)sin(fmod(0.5f*ph,TwoPi)))-1.f;

// delete casts, it's all floats anyways
case 1: w = fabs(2.f * sin(fmod(0.5f * ph, TwoPi))) - 1.f;

// remove unnecessary fmod(), sin() is well-defined for inputs outside [0, τ]
case 1: w = fabs(2.f * sin(0.5f * ph)) - 1.f;

// convert c-style math functions to std::* c++ counterparts
case 1: w = std::abs(2.f * std::sin(0.5f * ph)) - 1.f;

// new code
if (form == 1) { return 2.f * std::abs(std::sin(0.5f * ph)) - 1.f; }
```

So my "new" version is in fact identical to the original, intended version, with the exception of the `2.f *` being outside the `abs()` rather than inside it. This was an unintentional change, but has no effect on the result in this case. I'll move it back inside to make it match the original better for a cleaner diff history though.

Reply:
Oh good catch!